src/Dialect/ONNX/Transforms/Recompose.cpp

```diff
@@ -602,6 +650,145 @@ struct RecomposeQLinearMatMulFromQuantizeLinearPattern
+      weightValues.push_back(gemm.getB());
+      totalOutputFeatures += weightType.getShape()[concatAxis];
+    }
+
+    weightDims[concatAxis] = totalOutputFeatures;
+    Type newWeightType = RankedTensorType::get(weightDims, elementType);
+    Value newWeight =
+        create.onnx.concat(newWeightType, weightValues, concatAxis);
+
+    // Concatenate biases (create zero constants for missing biases)
+    SmallVector<Value> biasValues;
+    for (auto gemm : parallelGemms) {
+      if (Value bias = gemm.getC()) {
+        biasValues.push_back(bias);
+      } else {
+        auto biasType =
+            RankedTensorType::get({totalOutputFeatures}, elementType);
+        Value zeroBias =
+            create.onnx.constant(DenseElementsAttr::get(biasType, 0.0));
+        biasValues.push_back(zeroBias);
+      }
+    }
+
+    SmallVector<int64_t> newBiasShape = {totalOutputFeatures};
+    Type newBiasType = RankedTensorType::get(newBiasShape, elementType);
+    Value newBias = create.onnx.concat(newBiasType, biasValues, 0);
+
+    // Create combined Gemm operation
+    auto outputShape =
+        mlir::cast<ShapedType>(gemmOp1.getResult().getType()).getShape().vec();
+    outputShape[Axis] = totalOutputFeatures;
+    auto newOutputType = RankedTensorType::get(outputShape, elementType);
+
+    auto newGemm = rewriter.create<ONNXGemmOp>(loc, newOutputType, input,
+        newWeight, newBias, gemmOp1.getAlphaAttr(), gemmOp1.getBetaAttr(),
+        gemmOp1.getTransAAttr(), gemmOp1.getTransBAttr());
+
+    // Check for common ConcatOp
+    ONNXConcatOp commonConcatOp = nullptr;
+    for (auto gemm : parallelGemms) {
+      for (auto user : gemm.getResult().getUsers()) {
+        if (auto concatOp = dyn_cast<ONNXConcatOp>(user)) {
+          if (!commonConcatOp) {
+            commonConcatOp = concatOp;
+          }
+          if (concatOp != commonConcatOp) {
+            commonConcatOp = nullptr;
+            break;
+          }
+        } else {
+          commonConcatOp = nullptr;
+          break;
+        }
+      }
+      if (!commonConcatOp) {
+        break;
+      }
+    }
+
+    if (commonConcatOp) {
+      commonConcatOp.getResult().replaceAllUsesWith(newGemm.getResult());
+      rewriter.eraseOp(commonConcatOp);
+    } else {
+      SmallVector<int64_t, 4> splitSizesVec;
+      for (auto gemm : parallelGemms) {
+        int64_t outputChannels =
+            mlir::cast<ShapedType>(gemm.getResult().getType()).getShape()[Axis];
+        splitSizesVec.push_back(outputChannels);
+      }
+
+      ArrayRef<int64_t> splitSizes(splitSizesVec);
+      ValueRange splitResults = onnx_mlir::emitSplitByChannels(
+          rewriter, loc, newGemm.getResult(), splitSizes, Axis);
+
+      for (size_t i = 0; i < parallelGemms.size(); ++i) {
+        parallelGemms[i].replaceAllUsesWith(splitResults[i]);
+      }
+    }
+
+    for (auto gemm : parallelGemms) {
+      rewriter.eraseOp(gemm);
```

Comment:
Did you check that these Gemm ops must have one use before erasing? Otherwise, the generated IR would be broken.